src/components/shared/MarkerTooltipContents.js

```diff
@@ -37,19 +39,80 @@ function _markerDetail<T: NotVoidOrNull>(
   fn: T => string = String
 ): React.Node {
   return [
-    <div className="tooltipLabel" key="{key}">
+    <div className="tooltipLabel" key={key}>
       {label}:
     </div>,
     fn(value),
   ];
 }
 
+type PhaseTimeTuple = { name: string, time: Microseconds };
+
+function _markerDetailPhase(p: PhaseTimeTuple): React.Node {
+  return _markerDetail(
+    'gcphase' + p.name,
+    'Phase ' + p.name,
+    p.time,
+    t => formatNumber(t / 1000) + 'ms'
+  );
+}
+
+function _makePhaseTimesArray(
+  phases: PhaseTimes<Microseconds>
+): Array<PhaseTimeTuple> {
+  const array = [];
+  for (const phase in phases) {
+    if (phase !== 'Total') {
+      array.push({ name: phase, time: phases[phase] });
+    }
+  }
+  return array;
+}
+
+function _leafPhase(phases: Array<PhaseTimeTuple>, name: string): boolean {
```

Comment:
From your perspective my old habits as a logic programmer stand out!  Each of your suggestions here was because I was used to the logic programming style ;-)